spring-cloud-dataflow-module-deployer-spi/src/main/java/org/springframework/cloud/dataflow/module/deployer/test/AbstractModuleDeployerTests.java

```diff
@@ -0,0 +1,277 @@
+import org.springframework.cloud.dataflow.core.ModuleDeploymentId;
+import org.springframework.cloud.dataflow.core.ModuleDeploymentRequest;
+import org.springframework.cloud.dataflow.module.ModuleStatus;
+import org.springframework.cloud.dataflow.module.deployer.ModuleDeployer;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+/**
+ * Abstract base class for integration tests of
+ * {@link org.springframework.cloud.dataflow.module.deployer.ModuleDeployer} implementations.
+ *
+ * <p>Inheritors should setup an environment with a newly created {@link ModuleDeployer} that has no pre-deployed
+ * modules. Tests in this class are independent and leave the deployer in a clean state after they successfully
+ * run.</p>
+ *
+ * <p>As deploying a module is often quite time consuming, some tests often test various aspects of deployment in a
+ * row, to avoid re-deploying modules over and over again.</p>
+ * @author Eric Bottard
+ */
+@RunWith(SpringJUnit4ClassRunner.class)
+public abstract class AbstractModuleDeployerTests {
+
+	@Autowired
+	protected ModuleDeployer moduleDeployer;
+
+	@Before
+	public void assertCleanStart() {
+		assertThat("A newly created deployer should report no modules " +
+						"(this error may be caused by a previous test failing and not cleaning after itself)",
+				moduleDeployer.status().isEmpty(), is(true));
+	}
+
+	@Test
+	public void testUnknownDeployment() {
+		ModuleDeploymentId id = new ModuleDeploymentId("agroup", "alabel");
+		ModuleStatus status = moduleDeployer.status(id);
+
+		assertThat(status.getModuleDeploymentId(), is(id));
+		assertThat("The map was not empty: " + status.getInstances(), status.getInstances().isEmpty(), is(true));
+		assertThat(status.getState(), is(unknown));
+	}
+
+	/**
+	 * Tests a simple deploy-undeploy cycle.
+	 */
+	@Test
+	public void testSimpleDeployment() {
+		ModuleDefinition definition = new ModuleDefinition.Builder()
+				.setGroup(randomName())
+				.setName(randomName())
+				.build();
+		ArtifactCoordinates coordinates = ArtifactCoordinates
+				.parse("org.springframework.cloud.stream.module:time-source:jar:exec:1.0.0.M2");
+		ModuleDeploymentRequest request = new ModuleDeploymentRequest(definition, coordinates);
+
+		ModuleDeploymentId deploymentId = moduleDeployer.deploy(request);
+		Attempts timeout = deploymentTimeout();
+		assertThat(deploymentId, eventually(hasStatusThat(Matchers.<ModuleStatus>hasProperty("state", is(deployed))), timeout.noAttempts, timeout.pause));
+
+		// Use this opportunity to also test the multi-status query
+		assertThat(moduleDeployer.status(), IsMapContaining.hasEntry(is(deploymentId),
+				Matchers.<ModuleStatus>hasProperty("state", is(deployed))));
+
+		timeout = undeploymentTimeout();
+		moduleDeployer.undeploy(deploymentId);
+		assertThat(deploymentId, eventually(hasStatusThat(Matchers.<ModuleStatus>hasProperty("state", is(unknown))), timeout.noAttempts, timeout.pause));
```

Comment:
No, it is indeed unknown. See discussion at https://github.com/spring-cloud/spring-cloud-dataflow/pull/293/files#r49177576